Dream house

Write a program that calculates how long it would take you to save enough money to pay the down payment for your dream house. You will ask the user for their annual salary, the percent of the salary they’d want to save, and the cost of their dream home.

You should assume that the down payment is 25% of the total cost of the home. The time should be expressed in months. Hint: annual salary / 12 will give you the monthly salary. 

In [3]:
curr_ann_sal = round(float(input("What's your annual salary?  ")),2)
percent_sal = round(float(input('What percent of your salary will you save? ')),2)
cost = round(float(input('How much does your dream home cost? ')),2)
sem_ann_raise = round(float(input("What's the semi-annual raise percentage? ")),2)
inv_ann_ret = round(float(input("What's the investment annual return? ")),2)
down_pay = round(cost* (25/100),2)

saving = 0
month = 1
while saving < down_pay:
    mon_sal = curr_ann_sal/12
    saving = saving+(mon_sal)*(percent_sal/100)  
    if saving >= down_pay:
        break
    month += 1

years = int(month/12)
mon = int(month%12)

print("You would have to work for %d months to afford the downpayment for your dream house. %d months or %d years and %d months"%(month,month,years,mon))


What's your annual salary?  120000
What percent of your salary will you save? 10
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 0
What's the investment annual return? 0
You would have to work for 250 months to afford the downpayment for your dream house. 250 months, 20 years and 10 months


Get a raise

Use your code from the previous part as a starting point.

Extending the calculator from the last question, lets assume (accurately so) that you get a raise about twice year (every six months). Keeping this in mind, prompt the user for a semi-annual raise percentage (which will apply every six months) For example, if you start at 100,000 and get a 5% raise every six months, your salary for the first six months would be 8,333.33/month(base $100,000), and 8,750/month (base $105,000) for the following six months, and so on…

Sample output 1:

What's your annual salary? 120000
What percent of your salary will you save? 10%
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 0
What's the investment annual return? 4

You would have to work for 183 months to afford the downpayment for your dream house.
183 months or 15 years and 3 months. 

Sample output 2:

What's your annual salary? 150000
What percent of your salary will you save? 10
How much does your dream home cost? 500000
What's the semi-annual raise percentage? 3%
What's the investment annual return? 4%

You would have to work for 75 months to afford the downpayment for your dream house.
75 months or 6 years and 3 months. 

Sample output 3:

What's your annual salary? 120000
What percent of your salary will you save? 10%
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 0
What's the investment annual return? 0%

You would have to work for 250 months to afford the downpayment for your dream house.
250 months or 20 years and 10 months. 

In [9]:
############################################################################################################

##Get a raise
# Get inputs for annual salary, percentage saved, cost of the house, semi annual raise and investment annal return from the user
curr_ann_sal = round(float(input("What's your annual salary?  ")),2)
percent_sal = float(input('What percent of your salary will you save? '))    
while percent_sal >100:
    percent_sal = round(float(input('What percent of your salary will you save? ')),2)
cost = round(float(input('How much does your dream home cost? ')),2)
sem_ann_raise = round(float(input("What's the semi-annual raise percentage? ")),2)
while sem_ann_raise <0:
    sem_ann_raise = round(float(input("What's the semi-annual raise percentage? ")),2)
inv_ann_ret = round(float(input("What's the investment annual return? ")),2)
###########################################################################################################

# Calculate downpayment from cost, monthly salary from annual salary and initialise saving as 0 and month to be 1
down_pay = round(cost* (25/100),2)
ann_sal = curr_ann_sal
mon_salary = ann_sal/12
saving = 0
month = 1
###########################################################################################################

# create a while loop to determine the number of months until savings is less than downpayment
while saving <= down_pay:
    
    if month % 6 == 1 and month != 1:                               #Loop to update the annual salary with semi annual raise
        curr_ann_sal = round(curr_ann_sal+ curr_ann_sal * (sem_ann_raise/100),2)
        mon_sal = round(curr_ann_sal/12,2)
        
        if inv_ann_ret != 0:                                        # Loop to check whether investment annual return exists or not
            
            if month % 12 == 1 and month!=1:                        # Loop to update investment annual return to savings
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving+(mon_sal)*(percent_sal/100),2)
                saving = round(saving+ saving*(inv_ann_ret/100),2)
                
            else:
                mon_sal = round(curr_ann_sal/12,2)  
                saving = round(saving+(mon_sal)*(percent_sal/100),2) # formula to calculate monthly savings 
                
        
        else:
            mon_sal = curr_ann_sal/12
            saving = saving+(mon_sal)*(percent_sal/100)
            if saving >= down_pay:
                break
            
    
    else:
        
        if inv_ann_ret != 0:
            
            if month % 12 == 1 and month != 1:
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving + (mon_sal)*(percent_sal/100),2)
                saving = round(saving +saving*(inv_ann_ret/100),2)
                
            else:
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving+(mon_sal)*(percent_sal/100),2)  
                        
        else:
            mon_sal = curr_ann_sal/12
            saving = saving+(mon_sal)*(percent_sal/100)
            if saving >= down_pay:
                break
    month+=1                                                           # month increment after calculation of monthly savings  
######################################################################################################            
           
years = int(month/12)                                                 # month to year and months conversion
mon = month%12
print("You would have to work for %d months to afford the downpayment for your dream house. %d months or %d years and %d months "%(month,month,years,mon))
####################################################################################################################


What's your annual salary?  150000
What percent of your salary will you save? 10
How much does your dream home cost? 500000
What's the semi-annual raise percentage? 3
What's the investment annual return? 4
You would have to work for 75 months to afford the downpayment for your dream house. 75 months or 6 years and 3 months 


Build some savings

You can reuse some of your code from the previous part.

The previous questions focused on how long you would have to work to afford the downpayment. In this question, you’re going to answer the reverse - how much would I have to save to be able to afford the downpayment in X years.

We can iron out some details prior, to make this easier: The semi-annual raise will be 7% The investments have an annual return of 4% The downpayment for the house will be 25% The dream house costs $1m.

For this question, you should find the best rate of savings so that the $1m downpayment can be paid in 3 years (36 months), given the user’s salary.

Sample run 1:

What's your annual salary? 120000
What percent of your salary will you save? 10%
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 0
What's the investment annual return? 4

You would have to work for 183 months to afford the downpayment for your dream house.
183 months or 15 years and 3 months. 

How many years are you planning to save for downpayment? 3
The best savings rate for you is 0.6548.

Sample run 2:

What's your annual salary? 150000
What percent of your salary will you save? 120
What percent of your salary will you save? 12%
How much does your dream home cost? 500000
What's the semi-annual raise percentage? -3%
What's the semi-annual raise percentage? 3%
What's the investment annual return? 7%

You would have to work for 62 months to afford the downpayment for your dream house.
62 months or 5 years and 2 months. 

How many years are you planning to save for downpayment? 4
The best savings rate for you is 0.1639.

Sample run 3:

What's your annual salary? 150000
What percent of your salary will you save? 10%
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 7%
What's the investment annual return? 4

You would have to work for 99 months to afford the downpayment for your dream house.
99 months or 8 years and 3 months. 

How many years are you planning to save for downpayment? 3
The best savings rate for you is 0.4411.

In [15]:
############################################################################################################

##Get a raise
# Get inputs for annual salary, percentage saved, cost of the house, semi annual raise and investment annal return from the user
curr_ann_sal = round(float(input("What's your annual salary?  ")),2)
percent_sal = float(input('What percent of your salary will you save? '))    
while percent_sal >100:
    percent_sal = round(float(input('What percent of your salary will you save? ')),2)
cost = round(float(input('How much does your dream home cost? ')),2)
sem_ann_raise = round(float(input("What's the semi-annual raise percentage? ")),2)
while sem_ann_raise <0:
    sem_ann_raise = round(float(input("What's the semi-annual raise percentage? ")),2)
inv_ann_ret = round(float(input("What's the investment annual return? ")),2)
###########################################################################################################

# Calculate downpayment from cost, monthly salary from annual salary and initialise saving as 0 and month to be 1
down_pay = round(cost* (25/100),2)
ann_sal = curr_ann_sal
mon_salary = ann_sal/12
saving = 0
month = 1
###########################################################################################################

# create a while loop to determine the number of months until savings is less than downpayment
while saving <= down_pay:
    
    if month % 6 == 1 and month != 1:                               #Loop to update the annual salary with semi annual raise
        curr_ann_sal = round(curr_ann_sal+ curr_ann_sal * (sem_ann_raise/100),2)
        mon_sal = round(curr_ann_sal/12,2)
        
        if inv_ann_ret != 0:                                        # Loop to check whether investment annual return exists or not
            
            if month % 12 == 1 and month!=1:                        # Loop to update investment annual return to savings
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving+(mon_sal)*(percent_sal/100),2)
                saving = round(saving+ saving*(inv_ann_ret/100),2)
                
            else:
                mon_sal = round(curr_ann_sal/12,2)  
                saving = round(saving+(mon_sal)*(percent_sal/100),2) # formula to calculate monthly savings 
                
        
        else:
            mon_sal = curr_ann_sal/12
            saving = saving+(mon_sal)*(percent_sal/100)
            if saving >= down_pay:
                break
            
    
    else:
        
        if inv_ann_ret != 0:
            
            if month % 12 == 1 and month != 1:
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving + (mon_sal)*(percent_sal/100),2)
                saving = round(saving +saving*(inv_ann_ret/100),2)
                
            else:
                mon_sal = round(curr_ann_sal/12,2)
                saving = round(saving+(mon_sal)*(percent_sal/100),2)  
                        
        else:
            mon_sal = curr_ann_sal/12
            saving = saving+(mon_sal)*(percent_sal/100)
            if saving >= down_pay:
                break
    month+=1                                                           # month increment after calculation of monthly savings  
######################################################################################################            
           
years = int(month/12)                                                 # month to year and months conversion
mon = month%12
print("You would have to work for %d months to afford the downpayment for your dream house. %d months or %d years and %d months "%(month,month,years,mon))
####################################################################################################################

## Build Some savings
# Get the number of years input from user and convert it into months
given_years = float(input("How many years are you planning to save for downpayment? "))
given_months  = int(given_years * 12)

######################################################################################################
#Initialise savings to be zero and saving rate will be between 0 and 1, To start of with, we take it to be 0.5  
steps = 0
curr_savings = 0
max_rate = float(1)
min_rate = float(0)
saving_rate = (max_rate+min_rate)/2.0
tolerance = down_pay/100                                               # this is quite high tolerance level which can be changed.

def calculateSavings(curr_savings, mon_salary, saving_rate, mons):     # function to calculate current savings
    for mons in range(0,given_months+1):
        if mons%6 ==1 and mons >1:                                     # Loop to calculate semi annual raise
            mon_salary = round(mon_salary * (1+(sem_ann_raise/100)),4) 
        curr_savings = round(curr_savings + mon_salary * saving_rate,4)  
    curr_savings = round(curr_savings * (1+ inv_ann_ret/100),4)        
    return(curr_savings)

while abs(curr_savings - down_pay) >= tolerance:                        # Loop to find out best saving rate
    curr_savings = calculateSavings(curr_savings, mon_salary, saving_rate, 1)
    if curr_savings <= down_pay:
        min_rate = saving_rate
        curr_savings = 0
    elif curr_savings > down_pay + tolerance:
        max_rate = saving_rate
        curr_savings = 0
    if (steps>100):                                                # I chose a maximum step number of 100 because my tolerance is high
        print("It is not possible to pay the down payment in three years")
        break
    saving_rate = (max_rate+min_rate)/2.0
    steps = steps+1
print("The best savings rate for you is ", round(saving_rate,4))

What's your annual salary?  150000
What percent of your salary will you save? 10
How much does your dream home cost? 1000000
What's the semi-annual raise percentage? 7
What's the investment annual return? 4
You would have to work for 99 months to afford the downpayment for your dream house. 99 months or 8 years and 3 months 
How many years are you planning to save for downpayment? 3
The best savings rate for you is  0.4414
